36-vending-machine-monitor/firmware/vending-machine/vending-machine.ino

```diff
@@ -0,0 +1,1125 @@
+        debug.println();
+        // clear out any current columns
+        initializeDispensingColumns(0);
+
+        if (count > COLUMN_COUNT_MAX) {
+            count = COLUMN_COUNT_MAX;
+        }
+
+        // ensure all XSHUT pins are pulled low, in case col_count is less than the number of actual sensors
+        for (int i=0; i<COLUMN_COUNT_MAX; i++) {
+            uint8_t pin = COLUMN_PINS[i];
+            pinMode(pin, OUTPUT);
+            digitalWrite(pin, LOW);
+        }
+
+        // I2C addresses of the ToF sensors must be an even number for host to sensor communication. 
+        // The LSB is used to signal direction.
+        uint8_t i2cAddress = 0x54;          // start with one more than the default of 0x52. This ensures that
+                                            // should a sensor "forget" its address, it can be contacted at a known address that doesn't collide with other sensors.
+        uint8_t col = 0;
+        while (count --> 0) {
+            DispensingColumn* column = new DispensingColumn(Wire, i2cAddress, COLUMN_PINS[col], col+1);
+            column->setEnabled(usbPower);
+            i2cAddress += 2;
+            col++;
+            column->next = columns;
+            columns = column;
+        }
+    }
+    else {
+        for (DispensingColumn* column = columns; column; ) {
+            DispensingColumn* next = column->next;
+            delete column;
+            column = next;
+        }
+        columns = nullptr;
+    }
+}
+
+/**
+ * @brief Sets the power state of the app. When powered by battery, all dispensing columns
+ * are disabled, and tilt sensing is deactivated.
+ * 
+ * @param powered 
+ */
+void powerStateChanged(bool powered) {
+    // enable/disable aux.serial
+    auxSerial.setEnabled(powered);
+    tiltSensor.setEnabled(powered);
+
+    // disable/enable monitoring of all dispensing columns
+    for (DispensingColumn* column = columns; column != nullptr; column = column->next) {
+        column->setEnabled(powered);
+    }
+
+    powerAlert.setActive(!powered);
+    notecardSleep.setEnabled(!powered);
+}
+
+/**
+ * @brief Sends monitoring events for all the columns.
+ */
+void sendReport() {
+    if (!usbPower)
+        return;
+
+    for (DispensingColumn* column = columns; column; column = column->next) {
+        column->sendMonitoringEvent();
+    }
+}
+
+/**
+ * @brief Updates `notecardVoltage` and `usbPower` values by sending a `card.voltage` request.
+ * 
+ * @return true the request was successful
+ * @return false the request was not successful. Values were not updated.
+ */
+bool readNotecardVoltage() {
+    J* req = notecard.newRequest("card.voltage");
+    J* rsp = notecard.requestAndResponse(req);
+    if (rsp && !NoteResponseError(rsp)) {
+        double voltage = JGetNumber(rsp, "value");
+        if (voltage<0.1) // i.e 0
+            return false;
+        usbPower = JGetBool(rsp, "usb");
+        notecard.deleteResponse(rsp);
+    }
+    return rsp;
```

Comment:
This function returns a bool but rsp is a pointer. This probably works, but it's a little confusing to read.